### PR TITLE
[@types/arc4] New package for arc4

### PR DESCRIPTION
**Pull request by maxiicodes**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by hofmmaxi** [comment]
What exactly is wrong with the tsconfig?

**Comment by peterblazejewicz** [comment]
> What exactly is wrong with the tsconfig?

the lib version derives from default 'es6' to 'es5', which is why it's flagged for review afaik

File: types/arc4/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+export type ArcAlgorithm = "arc4" | "rc4+" | "rc4a" | "vmpc";
+
+type AlgoToClass<T extends ArcAlgorithm> = T extends "arc4" ? Arc4
+    : T extends "rc4+" ? Rc4p
+    : T extends "rc4a" ? Rc4a
+    : T extends "vmpc" ? Vmpc
+    : never;
+
+export default function arc4<T extends ArcAlgorithm>(
```
Comment (peterblazejewicz) [inline types/arc4/index.d.ts:11]: Not sure how this came into declaration. Module seems to be plain, classic CJS:
https://unpkg.com/browse/arc4@3.4.0/index.js
so more like (pseudocode) imo:
```ts
declare function arc4(): ...

declare namespace arc4 {
    function normal(): ...
    function lodash(): ...
}

export = arc4;
```
(include types in namespace to get them exported)

**Comment by hofmmaxi** [comment]
> > What exactly is wrong with the tsconfig?
> 
> the lib version derives from default 'es6' to 'es5', which is why it's flagged for review afaik

Ah, I see. Thanks!

File: types/arc4/index.d.ts (hunk 1 above)
Comment (hofmmaxi) [reply types/arc4/index.d.ts]: But how do I fix the compiler error now?

File: types/arc4/arc4-tests.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import arc4, { lodash, normal } from "arc4";
```
Comment (peterblazejewicz) [inline types/arc4/arc4-tests.ts:1]: you could try:
```ts
import arc4 = require("arc4");
const { lodash, normal } = arc4;
```
(that's only needed on DT, package can be consumed in common way with default TS configs)

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Comment by hofmmaxi** [comment]
Ready to merge

**Pull request merged**